model/convert/service_event.go

```diff
@@ -0,0 +1,326 @@
+package convert
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/encoding/json"
+
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/order"
+	"github.com/onflow/flow-go/module/signature"
+)
+
+// ServiceEvent converts a service event encoded as the generic flow.Event
+// type to a ServiceEvent type for use within protocol software and protocol
+// state. This acts as the conversion from the Cadence type to the flow-go type.
+func ServiceEvent(event flow.Event) (*flow.ServiceEvent, error) {
+
+	// depending on type of Epoch event construct Go type
+	switch event.Type {
+	case flow.EventEpochSetup:
+		return convertServiceEventEpochSetup(event)
+	case flow.EventEpochCommit:
+		return convertServiceEventEpochCommit(event)
+	default:
+		return nil, fmt.Errorf("invalid event type: %s", event.Type)
+	}
+}
+
+// convertServiceEventEpochSetup converts a service event encoded as the generic
+// flow.Event type to a ServiceEvent type for an EpochSetup event
+func convertServiceEventEpochSetup(event flow.Event) (*flow.ServiceEvent, error) {
+
+	// decode bytes using jsoncdc
+	payload, err := json.Decode(event.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal event payload: %w", err)
+	}
+
+	// parse cadence types to required fields
+	setup := new(flow.EpochSetup)
+
+	// NOTE: variable names prefixed with cdc represent cadence types
+	cdcEvent := payload.(cadence.Event)
+
+	// extract simple fields
+	setup.Counter = uint64(cdcEvent.Fields[0].(cadence.UInt64))
```

Comment:
Maybe check the field length here before accessing them, and conditionally cast, i.e. handle casting failures gracefully